### PR TITLE
Improvements/Cleanup of C++ Wrapper classes

### PR DESCRIPTION
**Pull request by gunterw**
### Description
The current C++ wrappers are ambiguous for certain data types and can't be used without explicit casting. The modifications remove the ambiguity and allow using the C++ wrappers based only on type of parameter passed to the functions.

### Test case
I have updated existing test cases to remove explicit casts and type fields when they are no longer necessary. 

I was not able to perform a make test for LUA bindings (did not work on my machine for some reason)

All other tests passed.

### Closure
There is not an open ticket for this change

### CODE REVIEW

File: swig/cpp/examples/cpp_turing_rpc_example.cpp [hunk 1]
```diff
@@ -56,9 +56,9 @@ class My_Callback:public sysrepo::Callback {
             auto output = holder->allocate(2);
 
             /* set 'output/step-count' leaf */
-            output->val(0)->set("/turing-machine:run-until/step-count", (uint64_t) 256, SR_UINT64_T);
+            output->val(0)->set("/turing-machine:run-until/step-count", (uint64_t) 256);
```
Comment (jktjkt) [inline swig/cpp/examples/cpp_turing_rpc_example.cpp]: Given that you're modifying these anyway, my recommendation is to use C++11-style construction of the numbers as well, i.e.:
```c++
foo->set("/some/xpath", uint64_t{123456});
```
Unlike the C-style cast which might be ambiguous in a C++ context (that's the reason why we have various `xxx_cast<>` statements), this one prevents errors when the value would overflow. These are examples, they should follow best practices.

File: swig/cpp/src/Struct.cpp [hunk 2]
```diff
@@ -88,7 +88,7 @@ char *Data::get_string() const {
 }
 uint8_t Data::get_uint8() const {
     if (_t != SR_UINT8_T) throw_exception(SR_ERR_DATA_MISSING);
-    return _d.uint32_val;
+    return _d.uint8_val;
```
Comment (jktjkt) [inline swig/cpp/src/Struct.cpp]: This looks like a genuine bugfix. I would be better done in an extra commit.

File: swig/cpp/src/Struct.cpp [hunk 3]
```diff
@@ -105,8 +105,7 @@ uint64_t Data::get_uint64() const {
 
 // Val
 Val::Val(sr_val_t *val, S_Deleter deleter) {
-    if (val == nullptr)
-        throw_exception(SR_ERR_INVAL_ARG);
+    if (val == nullptr) throw_exception(SR_ERR_INVAL_ARG);
```
Comment (jktjkt) [inline swig/cpp/src/Struct.cpp]: Why are these changes here? They do not appear to follow the coding style conventions as used in the rest of the code base...

Comment (gunterw) [reply swig/cpp/src/Struct.cpp]: I was getting compiler warnings. The if statement should either be a one liner, or it should have { }'s around the statement to eliminate the warning.

Which is preferred? 

if (vall == nullptr) throw_exception(SR_ERR_INVAL_ARG);

or 

if (vall == nullptr) {
   throw_exception(SR_ERR_INVAL_ARG);
}

Comment (gunterw) [reply swig/cpp/src/Struct.cpp]: After looking at the code some more, most of the existing if () throw_exception lines in the code are one liners, there are a few left that are still two liners without parens. There is nothing in the coding style prohibiting one liners, so I will convert the few remaining instances to one liners to be consistent

Comment (gunterw) [reply swig/cpp/src/Struct.cpp]: Actually looking at all of the files it's 50/50 so I will change to two liners with {}

File: swig/cpp/src/Struct.cpp [hunk 4]
```diff
@@ -247,161 +269,159 @@ Val::Val(int64_t int64_val, sr_type_t type) {
...
-        throw_exception(SR_ERR_INVAL_ARG);
     }
-
-    val->type = type;
+    val->data.uint64_val = uint64_val;
+    val->type = SR_UINT64_T;
     _val = val;
     _deleter = S_Deleter(new Deleter(val));
 }
 void Val::set(const char *xpath, const char *value, sr_type_t type) {
```
[the remark this answers is not in the document]
Comment (gunterw) [reply swig/cpp/src/Struct.cpp]: The set methods that still have the type specified are for the language bindings (lua, python).
I think the char*, bool, decimal64, and int64 methods are the minimum required for the language bindings. I was not sure about whether uint64 was needed as well, I'm not familiar with integer conversions in lua/python and whether a uint64 with MSB set would get lost so I left the type field on uint64 too.

File: swig/cpp/src/Struct.cpp [hunk 5]
```diff
@@ -140,87 +175,72 @@ Val::Val(const char *value, sr_type_t type) {
...
-        throw_exception(SR_ERR_INVAL_ARG);
     }
-
-    val->type = type;
+    val->data.int32_val = int32_val;
+    val->type = SR_INT32_T;
     _val = val;
     _deleter = S_Deleter(new Deleter(val));
 }
 Val::Val(int64_t int64_val, sr_type_t type) {
```
[the remark this answers is not in the document]
Comment (gunterw) [reply swig/cpp/src/Struct.cpp]: Can you be more specific? not sure which overload you are discussing and why it could be remove

File: swig/cpp/src/Struct.hpp [hunk 6]
```diff
@@ -103,63 +103,63 @@ class Val
      * SR_IDENTITYREF_T and SR_INSTANCEID_T.*/
     Val(const char *val, sr_type_t type = SR_STRING_T);
     /** Constructor for bool value.*/
-    Val(bool bool_val, sr_type_t type = SR_BOOL_T);
+    explicit Val(bool bool_val, sr_type_t type = SR_BOOL_T);
     /** Constructor for decimal64 value.*/
-    Val(double decimal64_val);
+    explicit Val(double decimal64_val, sr_type_t type = SR_DECIMAL64_T);
```
[the remark this answers is not in the document]
Comment (gunterw) [reply swig/cpp/src/Struct.hpp]: If this is not required for language bindings it could be removed

File: swig/cpp/src/Struct.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jktjkt) [reply swig/cpp/src/Struct.cpp]: I am referring to the constructor which takes an `int64_t` and at the same time also an explicit type specification. Of course, it is still important to be able to pass "just" the `int64_t` as-is, but this constructor also supports code like this:

```c++
auto foo = Val(10, SR_UINT8_T);
```

File: swig/cpp/src/Struct.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jktjkt) [reply swig/cpp/src/Struct.cpp]: @mislavn should know.

**Comment by gunterw** [comment]
FYI - after some issues are discussed/resolved on this pull request, I will close it and created a few separate pull requests

File: swig/cpp/src/Struct.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gunterw) [reply swig/cpp/src/Struct.cpp]: Again this is needed for language bindings, otherwise the type arg could be removed and you could do this instead.

`auto foo = Val(uint8{10})`

File: swig/cpp/src/Struct.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mislavn) [reply swig/cpp/src/Struct.cpp]: Lua and Python SWIG bindings can only see int64, so that's why in SWIG I ignore the rest of the functions because from their perspective they have the same function signature. Also that's why int64 contains logic for setting all other integers.

[the remark this answers is not in the document]
Comment (gunterw) [reply swig/cpp/src/Struct.cpp]: OK I'll remove explicit type args for uint64
